### PR TITLE
Add sprepost to superoperators

### PR DESCRIPTION
**Pull request by akirakyle**
`sprepost` is necessary for creating superoperators with different input and output hilbert space dimensions since `spre(A)*spost(B) != sprepost(A, B)` in such cases.

Note that I'm not sure exactly what the sensible behavior should be when calling `spre` or `spost` with an operator that doesn't have the same `basis_l` and `basis_r` and I've opened a relevant issue with qutip (https://github.com/qutip/qutip/issues/2171).

### CODE REVIEW

**Comment by Krastanov** [comment]
Please excuse the long delay on this, I wanted to find some time to read through your qutip issue.

Any chance you can add to the tests something that verifies the `spre(A)*spost(B) != sprepost(A, B)` case you mentioned, just so that if someone in the future decides to do some optimizations, they do not inadvertently break this edge case? Otherwise it looks good to merge.

File: test/test_apply.jl [hunk 1]
```diff
@@ -36,19 +36,4 @@ st = coherentstate(FockBasis(1), 1.4)
...
-k1 = coherentstate(b1, 0.39)
-k2 = coherentstate(b2, 1.4)
-op = projector(k1, k2')
-sOp2 = spre(op)
-
-try
-    QuantumOpticsBase.apply!(st, 1, sOp2)
-catch e
-    @test typeof(e) <: QuantumInterface.IncompatibleBases
-end
```
Comment (Krastanov) [inline test/test_apply.jl:52]: Any reason to remove this test? Should it not simply test for the new error you have?

Comment (akirakyle) [reply test/test_apply.jl:52]: The test wouldn't be for `apply!` anymore since the error is now thrown when calling `spre(op)` which is already tested in `test_superoperators.jl` so I didn't think it made sense have such a test in this file.

Comment (Krastanov) [reply test/test_apply.jl:52]: Let's still keep them -- down the line something might be modified further, and a mistake might be caught only by these tests. Generally, avoid removing tests, even if they are redundant -- at some point they might not have been redundant and we do not want to risk further refactoring in the future making them necessary again.

Comment (akirakyle) [reply test/test_apply.jl:52]: I just reworked this test to now pass but without removing it. I left the previously tested for error commented out in the spirit of avoiding removing tests.

**Comment by Krastanov** [comment]
Thank you for implementing and polishing this!
